File: src/auth/auth.service.ts
import { BadRequestException, Injectable } from '@nestjs/common';
import { CreateUserDto } from './dtos/create-user.dto';
import { LoginDto } from './dtos/login.dto';
import { UsersService } from 'src/users/users.service';
import * as bcrypt from 'bcrypt';

@Injectable()
export class AuthService {
  constructor(private usersService: UsersService) {}
  register(dto: CreateUserDto) {}

  async login(dto: LoginDto) {
    const user = await this.usersService.findByEmail(dto.username);
    if (!user || !bcrypt.compareSync(dto.password, user.password)) {
      throw new BadRequestException({
        errors: { message: 'incorrect email or password' },
      });
    }
    return user;
  }
}

File: internal/service/auth_service.go
package service

import (
	"errors"
	"fmt"

	"github.com/usmaarn/yummeals_api/internal/dto"
	"github.com/usmaarn/yummeals_api/internal/model"
	"github.com/usmaarn/yummeals_api/pkg/utils/constants"
	"golang.org/x/crypto/bcrypt"
)

type AuthService struct {
	userService  *UserService
	tokenService *TokenService
}

func NewAuthService(userService *UserService, tokenService *TokenService) *AuthService {
	return &AuthService{userService, tokenService}
}

func (s *AuthService) Register(dto *dto.CreateUserRequest) (*model.Token, error) {
	if dto == nil {
		return nil, fmt.Errorf("user data cannot be nil")
	}
	user, err := s.userService.CreateUser(dto)
	if err != nil {
		return nil, err
	}

	return s.tokenService.CreateToken(user.ID)
}
func (s *AuthService) Login(dto dto.LoginUserRequest) (*model.Token, error) {
	user, err := s.userService.FindByEmail(dto.Email)
	if err != nil {
		return nil, errors.New("invalid email or password")
	}

	err = bcrypt.CompareHashAndPassword([]byte(user.Password), []byte(dto.Password))
	if err != nil {
		return nil, errors.New("invalid email or password")
	}

	if user.Status != constants.UserStatusActive {
		return nil, fmt.Errorf("user is not active")
	}

	return s.tokenService.CreateToken(user.ID)
}
